scylla/src/transport/load_balancing/mod.rs

```diff
@@ -19,13 +20,89 @@ pub struct StatementInfo<'a> {
     pub keyspace: Option<&'a str>,
 }
 
-pub type Plan<'a> = Box<dyn Iterator<Item = Arc<Node>> + Send + Sync + 'a>;
-
 /// Policy that decides which nodes to contact for each query
 pub trait LoadBalancingPolicy: Send + Sync {
-    /// It is used for each query to find which nodes to query first
-    fn plan<'a>(&self, statement_info: &StatementInfo, cluster: &'a ClusterData) -> Plan<'a>;
+    /// It is used for each query to find which nodes to query.\
+    /// The generated plan can borrow any of the arguments for the whole query.
+    fn plan<'a>(
+        &'a self,
+        statement_info: &'a StatementInfo,
+        cluster: &'a ClusterData,
+    ) -> LBPlan<'a>;
+
+    /// Returns name of this load balancing policy
+    fn name(&self) -> Cow<'_, str>;
+}
+
+/// Represents a load balancing plan generated by a load balancing policy.\
+/// Common policies have their plan added as an enum to avoid needless allocations.\
+/// In case of a custom policy it has to allocate a Box with its plan inside.
+pub enum LBPlan<'a> {
+    Dumb(DumbPlan<'a>),
+    Custom(Box<dyn LoadBalancingPlan<'a> + Send + 'a>),
+}
+
+/// Each load balancing plan implements this trait.\
+/// A plan is almost like an Iterator<Item = &Arc<Node>> but the lifetimes must be correct.\
+pub trait LoadBalancingPlan<'a> {
+    fn next(&mut self) -> Option<&'a Arc<Node>>;
+}
+
+impl<'a> LoadBalancingPlan<'a> for LBPlan<'a> {
+    fn next(&mut self) -> Option<&'a Arc<Node>> {
+        match self {
+            LBPlan::Dumb(dumb_plan) => dumb_plan.next(),
+            LBPlan::Custom(custom_plan) => custom_plan.next(),
+        }
+    }
+}
+
+/// Sometimes a load balancing plan needs to be shared between multiple threads.
+/// This happens e.g. in case of speculative execution where multiple query attempts
+/// try to use consecutive nodes from a single plan. This is a wrapper over LBPlan which
+/// converts it into a thread safe version. Then LoadBalancingPlan is implemented
+/// for &SharedLBPlan which can be passed to other functions like any other plan.
+/// TODO: It could be implemented in a more efficient way, for example it could use
+/// AtomicUsize in round robin instead of using Mutex.
+pub struct SharedLBPlan<'a> {
+    plan: Mutex<LBPlan<'a>>,
+}
+
+impl<'a> From<LBPlan<'a>> for SharedLBPlan<'a> {
+    fn from(plan: LBPlan<'a>) -> SharedLBPlan<'a> {
+        SharedLBPlan {
+            plan: Mutex::new(plan),
+        }
+    }
+}
+
+impl<'a> LoadBalancingPlan<'a> for &SharedLBPlan<'a> {
+    fn next(&mut self) -> Option<&'a Arc<Node>> {
+        let mut_ref = &mut self.plan.lock().unwrap();
```

Comment:
`unwrap()` in library code deserves an explanation why it's always true; and if it isn't, it should be some kind of a runtime error instead